gamla/transducer.py

```diff
@@ -44,22 +70,47 @@ def _replace_index(the_tuple, index, value):
 
 
 def map(f: Callable[[Any], Any]):
+    """
+    Transducer version of `map`.
+
+    >>> transducer.transduce(
+        transducer.map(lambda x: x + 1),
+        lambda previous, current: (*previous, current),
+        (),
+        [1, 2, 3],
+    )
+    (3, 4, 5)
+    """
     return lambda step: lambda s, current: step(s, f(current))
 
 
 def filter(f: Callable[[Any], bool]):
+    """
+    Transducer version of `filter`.
+
+    >>> transducer.transduce(
+        transducer.filter(lambda x: x > 2),
+        lambda previous, current: (*previous, current),
+        [],
+        [1, 2, 3]
+    )
+    (3,)
+    """
     return lambda step: lambda s, x: step(s, x) if f(x) else s
 
 
 def concat(step: Reducer):
+    """Flattens collections using the `step` function into a single collection."""
```

Comment:
is it the same behaviour as functional_generic.concat? if it is, is there any difference/reason to use transducer? if it isn't, i think an example will be good

Reply:
yeah it's different, but it's more like an entire talk why it's different, and relevant to all the functions in this namespace. The users of `transducer` are asumed to have this knowledge.

The basic idea is that it doesn't assume anything about the collection (so it's not a generator, for example). One benefit is not having to worry about generator depletion.

wdyt?